scrapers/nd/events.py

```diff
@@ -1,114 +1,174 @@
+    def __init__(self, items, url):
+        self.items = items
+        self.events = {}
+        self._tz = pytz.timezone("US/Central")
+        self.url = url
+
+    def consolidate(self):
+        for item in self.items:
+            date_time = item["date_time"]
+            local_date = dateutil.parser.parse(date_time)
+
+            item_date, item_time = str(local_date).split()
+
+            committee = item["committee"]
+            location = item["location"]
+            event_key = f"{item_date}-{committee}-{location}"
+
+            if not self.events.get(event_key):
+                self.events[event_key] = {}
+                self.events[event_key]["event_start_time"] = item_time
+                self.events[event_key]["item_keys"] = set()
+            else:
+                current_start = self.events[event_key]["event_start_time"]
+                if time_is_earlier(item_time, current_start):
+                    self.events[event_key]["event_start_time"] = item_time
+
+            self.events[event_key]["item_keys"].add(item_time)
+
+            agenda_item_details = {
+                "description": item["description"],
+                "bill_name": item["bill_name"],
+                "sub_com": item["sub_com"],
+            }
+            self.events[event_key][item_time] = []
+            self.events[event_key][item_time].append(agenda_item_details)
+
+        yield from self.create_events()
+
+    def create_events(self):
+        event_collection = []
+
+        for event in self.events.keys():
+            date, com, loc = re.search(r"(.+)\-(.+)\-(.+)", event).groups()
+
+            start_time = self.events[event]["event_start_time"]
+            date_time = f"{date} {start_time}"
+            date_object = dateutil.parser.parse(date_time)
+            date_with_offset = self._tz.localize(date_object)
+
+            event_obj = Event(
+                name=com,
+                location_name=loc,
+                description="Standing Committee Hearing",
+                start_date=date_with_offset,
+            )
 
-                    event.add_document(
-                        link_text, link.xpath("@href")[0], media_type="application/pdf"
+            for item_key in self.events[event]["item_keys"]:
+                agenda = self.events[event][item_key]
+                for item in agenda:
+                    time = datetime.datetime.strptime(item_key, "%H:%M:%S").strftime(
+                        "%I:%M %p"
                     )
+                    descr_with_time = f"[{time}]: {item['description']}"
+                    item_descr = event_obj.add_agenda_item(descr_with_time)
+                    if item["bill_name"]:
+                        item_descr.add_bill(item["bill_name"])
+                    if item["sub_com"]:
+                        item_descr.extras["sub_committee"] = item["sub_com"]
+
+            event_obj.add_source(self.url)
+
+            event_collection.append(event_obj)
```

Comment:
Perhaps do a `yield event_obj` here (and remove the `event_collection` object entirely) so the upper generator can actually yield results as they come in rather than all at once?

Reply:
That makes sense, [and I changed it accordingly here](https://github.com/openstates/openstates-scrapers/pull/4162/commits/037209511528934a8e11f7616c83939e636b489a#diff-fa16689e28573ece469e8dfea222521c243ef203a1c8c974d48d6b050b3ed308L93-R93). However, I noticed the feedback in the terminal still doesn't reflect the creation of any new data files until the final moment of the scraper runtime. I'm assuming this is because it's the NDEventScraper(Scraper) object that finally initializes the json files when the scrape has completed? Or is it due to the way I'm yielding/returning objects in the current class construction anywhere in [the overall file](https://github.com/openstates/openstates-scrapers/pull/4162/commits/037209511528934a8e11f7616c83939e636b489a#diff-fa16689e28573ece469e8dfea222521c243ef203a1c8c974d48d6b050b3ed308)?

Reply:
I believe that's a result of how scrapelib/spatula generates the final json file(s), yes. This is more a "if you're treating the function as a generator, probably actually make it one" type of suggestion. I don't think it _really_ affects performance/etc. in any meaningful way.

Reply:
Okay, good to know.